synapse/federation/transport/client.py

```diff
@@ -148,6 +148,38 @@ async def backfill(
             destination, path=path, args=args, try_trailing_slash_on_400=True
         )
 
+    @log_function
+    async def timestamp_to_event(
+        self, destination: str, room_id: str, timestamp: int, direction: str
+    ) -> Optional[JsonDict]:
+        """
+        Calls a remote federating server at `destination` asking for their
+        closest event to the given timestamp in the given direction.
+
+        Args:
+            destination: Domain name of the remote homeserver
+            room_id: Room to fetch the event from
+            timestamp: The point in time we should navigate from in the given
+                direction to find the closest event.
+            direction: ["f"|"b"] to indicate whether we should navigate forward
+                or backward from the given timestamp to find the closest event.
+
+        Returns:
+            Results in a dict received from the remote homeserver.
+            Expected response will include the `event_id` key for the closest event.
+        """
+        path = _create_path(
+            FEDERATION_UNSTABLE_PREFIX,
+            "/org.matrix.msc3030/timestamp_to_event/%s",
+            room_id,
+        )
+
+        args = {"ts": [str(timestamp)], "dir": [direction]}
+
+        return await self.client.get_json(
+            destination, path=path, args=args, try_trailing_slash_on_400=True
+        )
```

Comment:
We ought to validate that the response is a `dict` and that `response["event_id"]` is a string, as expected.

Reply:
Also we'll never return `None` since the remote server will 404 if it can't find a suitable event.

Reply:
> We ought to validate that the response is a `dict` and that `response["event_id"]` is a string, as expected.

That seems reasonable except we don't do it for any of the other functions in this file AFAICT. Do we want to make this one different?

> Also we'll never return `None` since the remote server will 404 if it can't find a suitable event.

Another homeserver implementation could potentially return anything.

Reply:
>> We ought to validate that the response is a dict and that response["event_id"] is a string, as expected.
>
> That seems reasonable except we don't do it for any of the other functions in this file AFAICT. Do we want to make this one different?

There's an ongoing push to have more thorough mypy coverage and this file is a big pain point. The intent is to eventually have all the methods in this file perform validation, but we're not there yet. With that said, it'd be okay to omit the validation for now since someone will get around to adding it eventually.

>> Also we'll never return None since the remote server will 404 if it can't find a suitable event.
>
> Another homeserver implementation could potentially return anything.

True, but do note that `get_json` claims to return either a `JsonDict` or `list` and never `None`:
https://github.com/matrix-org/synapse/blob/develop/synapse/http/matrixfederationclient.py#L961-L970

(I suspect that `Union[JsonDict, list]` annotation is misleading and we're going to have a very bad time getting mypy to accept this file)

Reply:
> True, but do note that `get_json` claims to return either a `JsonDict` or `list` and never `None`:
https://github.com/matrix-org/synapse/blob/develop/synapse/http/matrixfederationclient.py#L961-L970

Ahh, I was looking at [`MatrixFederationRequest` which had `Optional[JsonDict]`](https://github.com/matrix-org/synapse/blob/e8ae94a22367a81049582dfdb16c743a45ca4e8b/synapse/http/matrixfederationclient.py#L179) 🤦‍♂️ (correct code is [under `MatrixFederationHttpClient` which @squahtx linked](https://github.com/matrix-org/synapse/blob/develop/synapse/http/matrixfederationclient.py#L961-L970))

I created `TimestampToEventResponse(TypedDict)` to better type what's being returned here